---
title: "Demonstrative workflow"
date: "02/16/2018"
output: 
     html_document
toc: true
fig_caption: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

The docker4seq package is an R control engine, which is at the core of the  SeqBox ecosystem.
It was developed to facilitate the use of computing demanding applications in the field of NGS data analysis.

The docker4seq package uses [docker](https://www.docker.com/what-docker) [containers](https://www.docker.com/what-container) that embed demanding computing tasks (e.g. short reads mapping) into isolated docker images. 

This approach provides multiple advantages: 

- user does not need to install all the software on its local server;

- docker images can be organized in pipelines;

- reproducible research is guarantee by the possibility of sharing the docker images used for the analysis. 

**MANDATORY:** The first time *docker4seq* is installed the **downloadContainers** function has to be executed to download, in the local repository, the docker images that are needed by *docker4seq*.

```{r, echo=TRUE, eval=FALSE}
library(docker4seq)
downloadContainers(group="docker")
```


## RNAseq workflow: Howto

The mRNAseq workflow can be run using **4SeqGUI** graphical interface (linux):

![mRNAseq workflow](/Users/raffaelecalogero/Dropbox/data/docker/stable/docker4seq/inst/img/rnaseq1.jpeg)

Sample quantification is made of these steps:

- Creating a genome index for STAR (see end of this paragraph)

- Running removing sequencing adapters

- Mapping reads to the reference genome

- Quantify gene and transcript expression level

- Annotating genes.

All the parameters can be setup using 4SeqGUI

\newpage

### Creating a STAR index file for mRNAseq:

The index can be easily created using the graphical interface:

![Creating a STAR genome index](/Users/raffaelecalogero/Dropbox/data/docker/stable/docker4seq/inst/img/rnaseq2.jpeg)

A detailed description of the parameters is given hereafter.

#### Creating a STAR index file by line command

\fontsize{8}{8}\selectfont
```{r,  echo=TRUE, eval=FALSE}
rsemstarIndex(group="docker",genome.folder="/data/scratch/hg38star",
ensembl.urlgenome="ftp://ftp.ensembl.org/pub/release-87/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.toplevel.fa.gz",
ensembl.urlgtf="ftp://ftp.ensembl.org/pub/release-87/gtf/homo_sapiens/Homo_sapiens.GRCh38.87.gtf.gz")
```

In brief, **rsemstarIndex** uses ENSEMBL genomic data. User has to provide the URL (**ensembl.urlgenome**) for the file XXXXX_dna.toplevel.fa.gz related to the organism of interest, the URL (**ensembl.urlgtf**) for the annotation GTF XXX.gtf.gz and the path to the folder where the index will be generated (**genome.folder**). The parameter **threads** indicate the number of cores dedicated to this task.

Precompiled index folders are available:

- [hg38star](http://130.192.119.59/public/hg38star.tar.gz)

- [mm10star](http://130.192.119.59/public/mm10star.tar.gz)

\newpage

### Quantifying genes/isoforms:

![Gene, Isoform counting](/Users/raffaelecalogero/Dropbox/data/docker/stable/docker4seq/inst/img/rnaseq3.jpeg)

A detailed description of the parameters is given below.

#### Sample quantification by line command

The sample quantification can be also executed using R and it is completely embedded in a single function:

```{r, echo=TRUE, eval=FALSE}

#test example
system("wget http://130.192.119.59/public/test.mrnaCounts.zip")
unzip("test.mrnaCounts.zip")
setwd("./test.mrnaCounts")
library(docker4seq)
rnaseqCounts(group="docker",fastq.folder=getwd(), scratch.folder=getwd(),
adapter5="AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT",
adapter3="AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT",
seq.type="se", threads=8,  min.length=40,
genome.folder="/data/scratch/mm10star", strandness="none", save.bam=FALSE,
org="mm10", annotation.type="gtfENSEMBL")

```


User needs to create the **fastq.folder**, where the fastq.gz file(s) for the sample under analysis are located. 
The **scratch.folder** is the location where temporary data are created. The results will be then saved in the **fastq.folder**.

User needs to provide also the sequence of the sequencing adapters, **adapter5** and **adapter3** parameters. 

**seq.type** indicates if single-end (se) or pair-end (pe) data are provided, **threads** indicates the max number of cores used by *skewer* and *STAR*, all the other steps are done on a single core.

The **min.length** refers to the minimal length that reads should have after adapters trimming. Since today the average read length for a RNAseq experiment is 50 or 75 nts then it would be better to bring to 40 nts the min.length parameter to increase the precision in assigning the correct position on the genome.

The **genome.folder** parameter refers to the location of the genomic index generated by STAR using the *docker4seq* function **rsemstarIndex**, see above paragraph.


**strandness**, is a parameter referring to the kit used for the library prep. If the kit does not provide strand information it is set to "none", if provides strand information is set to "forward" for  Illumina stranded kit and it set to "reverse" for Illumina ACCESS kit. **save.bam** set to TRUE indicates that genomic bam file and transcriotomic bam files are also saved at the end of the analysis. **annotation.type** refers to the type of available gene-level annotation. At the present time is only available ENSEMBL annotation defined by the gtf downloaded during the creation of the indexed genome files, see paragraph *at the end*Creating a STAR index file for mRNAseq*.


### Sample quantification output files

The mRNAseq workflow produces the following output files: 

    + XXXXX-trimmed.log, containing the information related to the adapters trimming
    + gtf_annotated_genes.results, the output of RSEM gene quantification with gene-level annotation
    + Log.final.out, the statistics of the genome mapping generated by STAR  
    + rsem.info, summary of the parameters used in the run
    + genes.results, the output of RSEM gene quantification
    + isoforms.results, the output of RSEM isoform quantification
    + run.info, some statistics on the run
    + skewerd_xxxxxxxxxxxx.log, log of the skewer docker container
    + stard.yyyyyyyyyyyy.log, log of the star docker container

![gtf_annotated_genes.results](/Users/raffaelecalogero/Dropbox/data/docker/stable/docker4seq/inst/img/rnaseq7.jpeg)

The first column in **gtf_annotated_genes.results** is the ensembl gene id, the second column is the [biotype](http://www.ensembl.org/Help/Faq?id=468), 
the third column is the annotation source, the fourth column contains the set of transcripts included in the ensembl gene id. Then there is the length of the gene, the length of the gene to which is subtracted the average length of the sequenced fragments, the expected counts are the counts to be used for differential expression analysis. [TPM](https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/) and [FPM](https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/) are normalized gene quantities to be used only for visualization purposes.


### Alignment-dependent tools versus aligner-free methods

Recently Zhang and coworkers (BMC Genomics 2017, 18,583) compared, at transcript level, alignment-dependent tools (Salmon_aln, eXpress, RSEM and TIGAR2) and aligner-free methods (Salmon, Kallisto Sailfish). In their paper, STAR was used as mapping tool for all alignment-dependent tools. In terms of isoform quantification, the authors indicated that there is strong concordance among quantification results from RSEM, Salmon, Salmon_aln, Kallisto and Sailfish (R2 > 0.89), suggesting that the impact of mappers on isoform quantification is small. Furthermore, the paper of Teng and coworkers (Genome Biology 2016, 17,74) reported that,in term of gene-level quantification, differences between alignment-dependent tools and aligner-free methods are shrinking with respect to transcripts level analysis.
On the basis of the above papers it seems that from the quantification point of view the difference between alignment free and alignment-dependent tools is very limited. 
However, aligner-free methods have low memory requirements and we have added Salmon in the development version of docker4seq in github. We are planning to introduce Salmon in the stable version of docker4seq in the first quarter of 2018. Salmon implementation will allow to increase the sample throughput, by running multiple samples. Currently samples are run serially because of the high RAM requirement of STAR.



### From samples to experiment

The RSEM output is sample specific, thus it is necessary to assemble the single sample in an experiment table including in the header of the columns both the covariates and the batches, if any.
The header sample name is separated by the covariate with an underscore, e.g. mysample1\_Cov1, mysample2\_Cov2. 

![counts table with covariates](/Users/raffaelecalogero/Dropbox/data/docker/stable/docker4seq/inst/img/counts1.jpeg)

A batch can be added to the sample name through a further underscore, e.g. mysample1\_Cov1\_batch1, mysample2\_Cov\_batch2.

![counts table with covariates and batch](/Users/raffaelecalogero/Dropbox/data/docker/stable/docker4seq/inst/img/counts2.jpeg)

The addition of the covariates to the various samples can be done using the **4seqGUI** using the button: *From samples to experiment*.

![generating a table with covariates](/Users/raffaelecalogero/Dropbox/data/docker/stable/docker4seq/inst/img/counts3.jpeg)

#### From samples to experiments by line command

```{r, echo=TRUE, eval=FALSE}
#test example
system("wget http://130.192.119.59/public/test.samples2experiment.zip")
unzip("test.samples2experiment.zip")
setwd("test.samples2experiment")
library(docker4seq)
sample2experiment(sample.folders=c("./e1g","./e2g","./e3g",
"./p1g", "./p2g", "./p3g"),
covariates=c("Cov.1","Cov.1","Cov.1","Cov.2","Cov.2","Cov.2"),
bio.type="protein_coding", output.prefix=".")

```

User needs to provide the paths of the samples, **sample.folder** parameter, a vector of the covariates, **covariates**, and the biotype(s) of interest, **bio.type** parameter. The parameter **output.prefix** refers to the path where the output will be created, as default this is the current R working folder.

#### From samples to experiments output files

This task produces the following output files: 

    + _counts.txt: gene-level raw counts table for differential expression analysis
    + _isoforms_counts.txt: isoform-level raw counts table for differential expression analysis
    + _isoforms_log2TPM.txt: isoform-level log2TPM for visualization purposes
    + _log2TPM.txt: gene-level log2TPM for visualization purposes
    + _isoforms_log2FPKM.txt: isoform-level log2FPKM for visualization purposes
    + _log2FPKM.txt: gene-level log2FPKM for visualization purposes
    + XXXXX.Rout: logs of the execution



### Visualizing experiment data with PCA

[PCA](https://georgemdallas.wordpress.com/2013/10/30/principal-component-analysis-4-dummies-eigenvectors-eigenvalues-and-dimension-reduction/) is a statistical procedure that uses an orthogonal transformation to convert a set of observations of possibly correlated variables into a set of values of linearly uncorrelated variables called principal components. This transformation is defined in such a way that the first principal component accounts for as much of the variability in the data as possible, and each succeeding component in turn has the highest variance possible under the constraint that it is orthogonal to the preceding components. 
**4SeqGUI** provides an interface to the generation experiment samples PCA

![PCA](/Users/raffaelecalogero/Dropbox/data/docker/stable/docker4seq/inst/img/pca1.jpeg)

The plot is saved in **pca.pdf** in the selected folder.

#### PCA by line command

```{r, echo=TRUE, eval=FALSE}
#test example
system("wget 130.192.119.59/public/test.analysis.zip")
unzip("test.analysis.zip")
setwd("test.analysis")
library(docker4seq)
pca(experiment.table="_log2FPKM.txt", type="FPKM", legend.position="topleft", 
    covariatesInNames=FALSE, principal.components=c(1,2), pdf = TRUE, output.folder=getwd())

```

User needs to provide the paths of experiment table, **experiment.table** parameter, i.e. the file generated using the samples2experiment function. The **type** parameter indicates if FPKM, TPM or counts are used by the PCA generation. The parameter **legend.position** defines where to locate the covariates legend. The parameter **covariatesInNames** indicates if the header of the experiment table contains or not covariate information. The parameter **principal.components** indicates which principal components should be plotted. **output.folder** indicates where to save the pca.pdf file.



![pca.pdf](/Users/raffaelecalogero/Dropbox/data/docker/stable/docker4seq/inst/img/pca2.jpeg)

The values in parentesis on x and y axes are the amount of variance explained by each principal component.

IMPORTANT: The above analysis is suitable for miRNAseq data too.



### Evaluating sample size and experiment power

Sample size estimation is an important issue in the design of RNA sequencing experiments. Furthermore, experiment power provides an indication of which is the fraction of differentially expressed genes that can be detected given a specific number of samples and differential expression detection thresholds.
[RnaSeqSampleSize](https://www.ncbi.nlm.nih.gov/pubmed/25246651) Bioconductor package provides the possibility to calculate, from a pilot experiment, the statistical power and to define the optimal sample size.
We have implemented wrapper functions to call these RnaSeqSampleSize functions for the sample size estimation and for statistical power estimation.


**4SeqGUI** provides an interface to sample size estimation and to statistical power estimation. 

![sample size estimation](/Users/raffaelecalogero/Dropbox/data/docker/stable/docker4seq/inst/img/ss.jpeg)



![stat power estimation](/Users/raffaelecalogero/Dropbox/data/docker/stable/docker4seq/inst/img/es.jpeg)


#### Sample size estimation by line command

```{r, echo=TRUE, eval=FALSE}
#test example
system("wget 130.192.119.59/public/test.analysis.zip")
unzip("test.analysis.zip")
setwd("test.analysis")
library(docker4seq)
sampleSize(group="docker", filename="_counts.txt", power=0.80, FDR=0.1, genes4dispersion=200, log2fold.change=1)

```

The requested parameters are the path of the counts experiment table generated by **samples2experiment** function. The param **power** indicates the expected fraction of differentially expressed gene, e.g 0.80. **FDR** and **log2fold.change**  are the two thresholds used to define the set of differentially expressed genes of interest. 

The output file is **sample_size_evaluation.txt** and it is saved in the R working folder, below an example of the file content:

![sample_size_evaluation.txt](/Users/raffaelecalogero/Dropbox/data/docker/stable/docker4seq/inst/img/samplesize1.jpeg)

IMPORTANT: The above analysis is suitable  for miRNAseq data too.



#### Experiment statistical power estimation by line command

```{r, echo=TRUE, eval=FALSE}
#test example
system("wget 130.192.119.59/public/test.analysis.zip")
unzip("test.analysis.zip")
setwd("test.analysis")
library(docker4seq)
experimentPower(group="docker", filename="_counts.txt",replicatesXgroup=7, FDR=0.1, genes4dispersion=200, log2fold.change=1)

```


The requested parameters are the path of the counts experiment table generated by **samples2experiment** function. The param **replicatesXgroup** indicates the number of sample associated with each of the two covariates. **FDR** and **log2fold.change**  are the two thresholds used to define the set of differentially expressed genes of interest. **genes4dispersion** indicates the number of genes used in the estimation of read counts and dispersion distribution.

The output file is **power_estimation.txt** and it is saved in the R working folder, below an example of the file content: 

![power_estimation.txt](/Users/raffaelecalogero/Dropbox/data/docker/stable/docker4seq/inst/img/expp1.jpeg)

IMPORTANT: The above analysis is suitable  for miRNAseq data too.



### Differential expression analysis with DESeq2

A basic task in the analysis of count data from RNA-seq is the detection of differentially expressed genes.
**4SeqGUI** provides an interface to DESeq2 to simplify differential expression analysis.

![DESeq2](/Users/raffaelecalogero/Dropbox/data/docker/stable/docker4seq/inst/img/de1.jpeg)

The output files are:

**DEfull.txt** containing the full set of results generated by DESeq2

![DEfull.txt](/Users/raffaelecalogero/Dropbox/data/docker/stable/docker4seq/inst/img/de2.jpeg)


**DEfiltered_log2fc_X_fdr_Y.Y.txt** containing the set of differentially expressed genes passing the indicated thresholds

![DEfiltered_log2fc_1_fdr_0.1.txt](/Users/raffaelecalogero/Dropbox/data/docker/stable/docker4seq/inst/img/de3.jpeg)


**genes4david.txt** a file containing only the gene symbols to be used as input for [DAVID](https://david.ncifcrf.gov/) or [ENRICHR](http://amp.pharm.mssm.edu/Enrichr/)

log2normalized_counts.txt, log2 library size normalized counts, calculated by DESeq2, that can be used for visualization purposes.

#### DESeq2 by line command

```{r, echo=TRUE, eval=FALSE}
#test example
system("wget 130.192.119.59/public/test.analysis.zip")
unzip("test.analysis.zip")
setwd("test.analysis")
library(docker4seq)
wrapperDeseq2(output.folder=getwd(), group="docker", 
      experiment.table="_counts.txt", log2fc=1, fdr=0.1, 
      ref.covar="Cov.1", type="gene", batch=FALSE)
```


User has to provide experiment table, **experiment.table** param, i.e. the counts table generated with **samples2experiment** function, the thresholds for the differential expression analysis, **log2fc** and **fdr** params, the reference covariate, **ref.covar** param, i.e. the covariate that is used as reference for differential expression detection, the **type** param, which refers to the type of experiment table in use: *gene*, *isoform*, *mirna*, **batch** parameter that indicates, if it is set to **TRUE** that the header of the experiment table also contains the extra information for the batch effect (see above). 

IMPORTANT: the above analysis can be applied to miRNAseq data too.

#### Why DESeq2 was chosen as differential expression tool

Love and co-workers, in their paper on DESeq2 (Love at al. Genome Biol. 2014; 15, 550), showed that DESeq2 had comparable sensitivity to edgeR and voom. We introduced in our workflow DESeq2 because it has some specific features which increase the strength of the differential expression analysis, features that are not available in other tools. One of these features is, the Empirical Bayes shrinkage for fold-change estimation, which shrinks log fold change estimates toward zero. This feature reduces the noise due to low expressed genes, since shrinkage is stronger when the available information for a gene is low, which may be because the read counts are low, dispersion is high or there are few degrees of freedom. The other peculiar feature of DESeq2 is the identification of counts outliers, which might represent a source of false positives. Specifically, DESeq2 flags genes characterized by the presence of counts outliers, which are estimated with the standard outlier diagnostic Cookâ€™s distance (Love at al. Genome Biol. 2014; 15, 550).



## miRNAseq workflow: Howto 

The miRNAseq workflow can be run using **4SeqGUI** graphical interface:

![miRNAseq workflow](/Users/raffaelecalogero/Dropbox/data/docker/stable/docker4seq/inst/img/mirna1.jpeg)
  
  
The miRNAseq docker container executes the following steps:


![miRNAseq workflow](/Users/raffaelecalogero/Dropbox/data/docker/stable/docker4seq/inst/img/mirna3.jpeg)

The full workflow is described in [Cordero et al. Plos ONE 2012](http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0031630). In brief, fastq files are trimmed using [cutadapt](https://github.com/marcelm/cutadapt) and the trimmed reads are mapped on miRNA precursors, i.e. harpin.fa file, from [miRBase](http://www.mirbase.org/ftp.shtml) using [SHRIMP](http://compbio.cs.toronto.edu/shrimp/). Using the location of the mature miRNAs in the precursor, countOverlaps function, from the Bioconductor package GenomicRanges is used to quantify the reads mapping on mature miRNAs. 

\newpage

All the parameters needed to run the miRNAseq workflow can be setup using 4SeqGUI

![miRNAseq parameters](/Users/raffaelecalogero/Dropbox/data/docker/stable/docker4seq/inst/img/mirna2.jpeg)
 
 
A detailed description of the parameters is given below.

### miRNAseq workflow by line command

The miRNAseq workflow can be also executed using R and it is completely embedded in a unique function:

```{r, echo=TRUE, eval=FALSE}
#test example
system("wget 130.192.119.59/public/test.mirnaCounts.zip")
unzip("test.mirnaCounts.zip")
setwd("test.mirnaCounts")
library(docker4seq)
mirnaCounts(group="docker",fastq.folder=getwd(), scratch.folder="/data/scratch", 
            mirbase.id="hsa",download.status=FALSE, adapter.type="NEB", trimmed.fastq=FALSE)

```

User has to create the **fastq.folder**, where the fastq.gz files for all miRNAs under analysis are located. 
The **scratch.folder** is the location where temporary data are created. The results will be then saved in the **fastq.folder**.
Moreover, user has to provide  the identifier of the miRBase organism, e.g. **hsa** for Homo sapiens, **mmu** for Mus musculus.  If the **download.status** is set to FALSE, mirnaCounts uses miRBase release 21, if it is set to TRUE the lastest version of precursor and mature miRNAs will be downloaded from miRBase. Users need to provide the name of the producer of the miRNA library prep kit to identify which adapters need to be provided to cutadapt, **adapter.type** parameter. The available adapters are NEB and Illumina, but, upon request, we can add other adapters. Finally, if the **trimmed.fastq** is set to FALSE  then the trimmed fastq are not saved at the end of the analysis.


### miRNAseq workflow output files

The miRNAseq workflow produces the following output files: 

    + README: A file describing the content of the data folder
    + all.counts.txt: miRNAs raw counts, to be used for differential expression analysis
    + trimmimg.log: adapters trimming statistics
    + shrimp.log: mapping statistics
    + all.counts.Rda: miRNAs raw counts ready to be loaded in R.
    + analysis.log: logs of the full analysis pipeline
  

\newpage
  
### Adding covariates and batches to mirnaCounts output: all.counts.txt

**4SeqGUI** provides an interface to add covariates and batches to all.counts.txt

![miRNAseq covariates and batches](/Users/raffaelecalogero/Dropbox/data/docker/stable/docker4seq/inst/img/mirna_covars.jpeg)

The function **mirnaCovar** add to the header of all.counts.txt covariates and batches or covariates only.

```{r, echo=TRUE, eval=FALSE}
#test example
system("wget 130.192.119.59/public/test.mirna.analysis.zip")
unzip("test.mirna.analysis.zip")
setwd("test.mirna.analysis")
library(docker4seq)
mirnaCovar(experiment.folder=paste(getwd(), "all.counts.txt", sep="/"),
     covariates=c("Cov.1", "Cov.1", "Cov.1", "Cov.1", "Cov.1", "Cov.1", 
                  "Cov.2", "Cov.2", "Cov.2", "Cov.2", "Cov.2", "Cov.2"),
     batches=c("bath.1", "bath.1", "bath.2", "bath.2", "batch.1", "batch.1", 
               "batch.2", "batch.2","batch.1", "batch.1","bath.2", "bath.2"), output.folder=getwd())
```


The output of **mirnaCovar**, i.e. w\_covar\_batch\_all.counts.txt, is compliant with PCA, Sample size estimator, Experiment stat. power and DEseq2 analysis.

\newpage

## chipseq workflow: HowTo

The chipseq workflow can be ran using **4SeqGUI** graphical interface:

![ChIPseq workflow](/Users/raffaelecalogero/Dropbox/data/docker/stable/docker4seq/inst/img/chipseq0.jpeg)


The ChIPseq consists of two main steps:

- Creating a genome index for BWA (see end of this paragraph)

- Running MACS or SICER analysis

\newpage

### Creating a BWA index file for Chipseq:


The index can be easily created using the graphical interface:

![Creating a BWA index with Genome indexing BWA](/Users/raffaelecalogero/Dropbox/data/docker/stable/docker4seq/inst/img/chipseq1.jpeg)

```{r,  echo=TRUE, eval=FALSE}
bwaIndexUcsc(group="sudo",genome.folder="/sto2/data/scratch/mm10bwa", uscs.urlgenome=
"http://hgdownload.cse.ucsc.edu/goldenPath/mm10/bigZips/chromFa.tar.gz",
gatk=FALSE)
```

In brief, **bwaIndexUcsc** uses UCSC genomic data. User has to provide the URL (**uscs.urlgenome**) for the file chromFa.tar.gz related to the organism of interest and the path to the folder where the index will be generated (**genome.folder**). The parameter **gatk** has to be set to FALSE if it is not required for ChIPseq genomic index creation. 

Precompiled index folders are available:

- [mm10bwa](http://130.192.119.59/public/mm10bwa.tar.gz)

\newpage

### Calling peaks and annotating:

All the parameters needed to run MACS or SICER can be setup using 4SeqGUI

![MACS and SICER analysis](/Users/raffaelecalogero/Dropbox/data/docker/stable/docker4seq/inst/img/chipseq3.jpeg)

A detailed description of the parameters is given below.

### Chipseq workflow by line command

The chipseq workflow can be also executed using R and it is completely embedded in a unique function:

```{r, echo=TRUE, eval=FALSE}

system("wget 130.192.119.59/public/test.chipseqCounts.zip")
unzip("test.chipseqCounts.zip")
setwd("test.chipseqCounts")
library(docker4seq)
chipseqCounts(group = "docker", output.folder = "./prdm51.igg",
  mock.folder="./igg", test.folder="./prdm51", scratch.folder=getwd(),
  adapter5 = "AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT",
  adapter3 = "AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT",
  threads = 8, min.length = 30, genome.folder,
  mock.id = "igg", test.id = "tf", genome, read.size = 50,
  tool = "macs", macs.min.mfold = 10, macs.max.mfold = 30,
  macs.pval = "1e-5", sicer.wsize = 200, sicer.gsize = 200,
  sicer.fdr = 0.1, tss.distance = 0, max.upstream.distance = 10000,
  remove.duplicates = "N")
```

Specifically user needs to create three folders:

    + mock.folder, where the fastq.gz file for the control sample is located. 
      For control sample we refer to ChIP with IgG only or input DNA.
    + test.folder, where the fastq.gz file for the ChIP of the sample to be analysed.
    + output.folder, where the R script embedding the above script is located.

The **scratch.folder** can be the same as the **output.folder**. However, if the system has a high speed disk for temporary calculation, e.g. a SSD disk, the location of the scratch.folder on the SSD will reduce significantly the total execution time.

User needs to provide also the sequencing adapters, i.e. **adapter5** and **adapter3** parameters. In case of Illumina platform  the adapters sequences can be easily recovered [here](https://support.illumina.com/content/dam/illumina-support/documents/documentation/chemistry_documentation/experiment-design/illumina-adapter-sequences_1000000002694-01.pdf). 

**Threads** indicates the max number of cores used by *skewer* and *bwa*, all the other steps are done on a single core.
The **min.length** refers to the minimal length that a reads should have after adapters trimming. Since today the average read length for a ChIP experiment is 50 or 75 nts  we suggest  to bring to 40 nts the min.length parameter to increase the precision in assigning the correct position on the genome.

The **genome.folder** parameter refers to the location of the genomic index generated by bwa using the *docker4seq* function **bwaIndexUcsc**.

**mock.id** and **test.id** identify the type of sample and are assigned to the ID parameter in the RG field of the bam file.

**genome** is the parameter referring to the annotation used to associate ChIP peaks with genes. In the present implementation  hg38, hg19 for human and mm10 and mm9 for mouse annotations are available.

**read.size** is a parameter requested by MACS and SICER for their analysis. 
**macs.min.mfold**, **macs.max.mfold**, **macs.pval**  are the default parameters requested to peaks definition for more info please refer to the documentation of MACS 1.4.
**sicer.wsize**, **sicer.gsize**, **sicer.fdr** are the default parameters requested to  peaks definition for more info please refer to the documentation of SICER 1.1.
**IMPORTANT**: The optimal value for **sicer.gsize** in case of H3K4Me3 ChIP is 200 and in case of ChIP H3K27Me3 is 600.   

**tss.distance** and **max.upstream.distance** are parameters required by ChIPseqAnno, which is the Bioconductor package used to assign the peaks to specific genes. Specifically max.upstream.distance refers to the max distance in nts that allows the association of a peak with a specific gene.

**remove.duplicates** is the parameter that indicates if duplicates have to be removed or not. It has two options: **N** duplicates are not removed, **Y** duplicates are removed.


### Chipseq workflow output files

The chipseq workflow produces the following output files: 

    + README: A file describing the content of the data folder
    + mypeaks.xls: All detected peaks alongside the nearest gene and its annotation
    + mytreat.counts: The total reads count for the provided treatment file
    + mycontrol.counts: The total reads count for the provided control/background file
    + peak_report.xls: Aggregate information regarding the peak and their position relative to the nearest gene
    + chromosome_distribution.pdf: Barplot of the distribution of the peaks on the chromosomes
    + relative_position_distribution.pdf: Barplot of the distribution of the peaks positions relative to their nearest gene
    + peak_width_distribution.pdf: Histogram of the distribution of the width of the peaks
    + distance_from_nearest_gene_distribution.pdf: Histogram of the distribution of the distance of each peak from its nearest gene
    + cumulative_coverage_total.pdf: Cumulative normalized gene coverage
    + cumulative_coverage_chrN.pdf: Cumulative normalized gene coverage for the specific chromosome
    + mycontrol_sorted.bw: bigWig file for UCSC Genome Browser visualization
    + mytreat_sorted.bw: bigWig file for UCSC Genome Browser visualization
     

## Tutorials

### RNAseq workflow

[**Tutorial experiment downloadable here**](http://130.192.119.59/public/test.mrnaCounts_full.zip): 

    + Three replicates for two experimental conditions 
    
    + single-end mode sequencing 
    
    + 1 million reads for each sample

**Experiment description**: 

    + 4T1 mouse cell line grown in standard DMEM medium (e) is compared with the same cells grown in low attachment medium (p)

The following data are available for download:

- [Fastq files for 3 samples grown in standard DMEM medium (e) and for 3 samples grown in low attachment medium (p) to be used to calculate samples counts](http://130.192.119.59/public/test.mrnaCounts_full.zip). 

    + This data set allows running all the steps required to detect differentially expressed genes. The first step is the quantification of genes and isoforms via mapping reads to the reference genome via STAR and using this mapping information to quantify genes and transcripts using RSEM, i.e. *this section*. 
    
    + The RSEM counts table, generated for all samples, are combined in a unique table, see *From samples to experiment* section. This table is used for differential expression genes detection. For visualization purposes  log2 FPKM and log2 TPM tables are also generated, see *From samples to experiment* section. More info on the FPKM and TPM are available [here](http://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/). 
    
    + Subsequently the overall characteristics of the dataset can be explore via [PCA](https://www.youtube.com/watch?v=BfTMmoDFXyE), *Visualizing experiment data with PCA* section. 
    
    + It is also possible to evaluate which is the statistical power of the experiment, i.e. identifying the fraction of genes/transcripts that can be identify giving the statistical structure of the experiment, or identify the optimal number of samples required to detect differentially expressed genes. More info in *Evaluating sample size and experiment power* section. 
    
    + Differential expression can be then evaluate using the DESeq2 module, *Differential expression analysis with DESeq2*. 


- Furthermore, also intermediate results are provided:

    + [The counts tables generated by STAR+RSEM analysis to be used to assemble counts, TPM and FPKM experiment tables](http://130.192.119.59/public/test.samples2experiment.zip)

    + [The experiment tables that can be used to study samples organization via PCA or the statistical characteristics of the experiment, i.e. the statistical power and the optimal sample size of the experiment](http://130.192.119.59/public/test.analysis.zip)

    + The above data set can be used also to detect differentially expressed genes and isoforms using the DESeq2 module.

### miRNAseq workflow

[**Tutorial experiment downloadable here**](http://130.192.119.59/public/test.mirnaCounts.zip): 

- Six specimens for two experimental conditions
    
- single-end mode sequencing, 
    
- 1 million reads for each sample.

**Experiment description**: 

- Six blood circulating exosomes miRNA samples from healthy donors (hd) and six blood circulating exosomes miRNA samples from tumor patients (tum)
    
- from 1 to 3 hd and tum samples were harvested on day 1, from 5 to 6 hd and tum samples were harvested on day 2. Thus the data are require the addition of the batch effect in differential expression analysis.

The following data are available for download:

- [Fastq files for 6 miRNA samples from healthy donors (hd), and six miRNA samples from tumor patients (tum) to be used to calculate samples counts and organize them in a counts table](http://130.192.119.59/public/test.mrnaCounts_full.zip). 

    + This data set allows running all the steps required to detect differentially expressed genes. The first step is the quantification of mature annotated miRNA and the generation of counts table to be used for differential expression, i.e. *this section*. 
    
    + The covariates and batch effects can be added to the counts table (all.counts.txt), see *From samples to experiment* section. This table is used for differential expression genes detection.  
    
    + Subsequently the overall characteristics of the dataset can be explore via [PCA](https://www.youtube.com/watch?v=BfTMmoDFXyE), *Visualizing experiment data with PCA* section. 
    
    + It is also possible to evaluate which is the statistical power of the experiment, i.e. identifying the fraction of genes/transcripts that can be identify giving the statistical structure of the experiment, or identify the optimal number of samples required to detect differentially expressed genes. More info in *Evaluating sample size and experiment power* section. 
    
    + Differential expression can be then evaluate using the DESeq2 module, *Differential expression analysis with DESeq2*.

### ChIPseq workflow

[**Tutorial experiment downloadable here**](http://130.192.119.59/public/test.chipseqCounts.zip): 

    + Two ChIPseq one IgG control and an other Prdm5 TF moAb (mouse)
    
    + single-end mode sequencing, 
    
    + 1 million reads for each sample.

**Experiment description**: 

- PRDM family members are transcriptional regulators involved in tissue specific differentiation. PRDM5 has been reported to predominantly repress transcription, but a characterization of its molecular functions in a relevant biological context is lacking. Prdm5 controls both Collagen I transcription and fibrillogenesis by binding inside the Col1a1 gene body and maintaining RNA polymerase II occupancy (Galli et al. PLoS Genet. 2012,e1002711).
    
- The toy experiment is organized in three folders: i. one for IgG (igg, containing the igg pool down fastq); ii. one for Prdm5 (prdm5, containing the Prdm5 pool down fastq) and iii. the other for analysis output prdm5.igg, where MACS results will be located. The execution of the ChIPseq workflow using GUI or line command will provide the final annotated table of peaks (mypeaks.xls)
 
    
