<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ docker4seq</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">docker4seq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/docker4seq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/skeleton.html">Skeleton tutorial</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="CIRCexplorer2.html">CIRCexplorer2()</a></code> </p>
        </td>
        <td><p>Running CIRCexplorer2 parse command for circRNAs prediction from STAR or BWA alignment output</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="annoByGtfchipseq.html">annoByGtfchipseq()</a></code> </p>
        </td>
        <td><p>Annotating RSEM gene.results using ENSEMBL gtf and refGenome CRAN package</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="anovaLike.html">anovaLike()</a></code> </p>
        </td>
        <td><p>A function allowing the identification of differentially expressed genes if multiple groups are provided.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bowtie2.html">bowtie2()</a></code> </p>
        </td>
        <td><p>Running bowtie2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bowtie2Index.html">bowtie2Index()</a></code> </p>
        </td>
        <td><p>Generating bowtie2 genome index</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bwa.html">bwa()</a></code> </p>
        </td>
        <td><p>Running bwa, Li and Durbin Bioinformatics, 2009 Jul 15;25(14):1754-60</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bwaIndex.html">bwaIndex()</a></code> </p>
        </td>
        <td><p>Generating bwa genome index</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bwaIndexUcsc.html">bwaIndexUcsc()</a></code> </p>
        </td>
        <td><p>Generating bwa genome index for GATK variant call</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cdhit.html">cdhit()</a></code> </p>
        </td>
        <td><p>A function to execute CD-HIT</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="chipseq.html">chipseq()</a></code> </p>
        </td>
        <td><p>Detection and annotation of TF binding sites and histone marks, based on the workflow used in Galli et al Mol Cell. 2015 Oct 15;60(2):328-37</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="chipseqCounts.html">chipseqCounts()</a></code> </p>
        </td>
        <td><p>Running MACS &amp; SICER workflow NOT READY FOR STABLE check it!</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="circrnaAnnotations.html">circrnaAnnotations()</a></code> </p>
        </td>
        <td><p>Annotation of a list of circRNAs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="circrnaBSJunctions.html">circrnaBSJunctions()</a></code> </p>
        </td>
        <td><p>Running CircHunter circRNA backsplicing sequence reconstruction module</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="circrnaClassification.html">circrnaClassification()</a></code> </p>
        </td>
        <td><p>Running CircHunter circRNA classification module</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="circrnaMergePredictions.html">circrnaMergePredictions()</a></code> </p>
        </td>
        <td><p>Function to merge different circRNA lists predicted from one of the supported circRNA prediction tools.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="circrnaOverlapResults.html">circrnaOverlapResults()</a></code> </p>
        </td>
        <td><p>Running circRNA prediction overlap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="circrnaPrepareFiles.html">circrnaPrepareFiles()</a></code> </p>
        </td>
        <td><p>Function to prepare the CircHunter reference annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="circrnaQuantification.html">circrnaQuantification()</a></code> </p>
        </td>
        <td><p>Running CircHunter circRNA quantification module</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="circrnaReformat.html">circrnaReformat()</a></code> </p>
        </td>
        <td><p>Running circRNA reformat script</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ciri2.html">ciri2()</a></code> </p>
        </td>
        <td><p>Running CIRI v2 tool for circRNAs prediction</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ciriAS.html">ciriAS()</a></code> </p>
        </td>
        <td><p>Running CIRI_AS tool for circRNAs structure prediction</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cutadapt.html">cutadapt()</a></code> </p>
        </td>
        <td><p>Function to execute Cutadapt on RNA-Seq reads</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="deeptoolBwig.html">deeptoolBwig()</a></code> </p>
        </td>
        <td><p>Create a bigWig from the output of bowtie2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="demultiplexing.html">demultiplexing()</a></code> </p>
        </td>
        <td><p>Generating running bcl2fastq</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dockerTest.html">dockerTest()</a></code> </p>
        </td>
        <td><p>Testing if Docker is installed</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="downloadContainers.html">downloadContainers()</a></code> </p>
        </td>
        <td><p>Download for the first time all containers embedded in the workflows</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="experimentPower.html">experimentPower()</a></code> </p>
        </td>
        <td><p>A wrapper function for experiment_power from RnaSeqSampleSize Bioconductor package</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fastq2fasta.html">fastq2fasta()</a></code> </p>
        </td>
        <td><p>A function to handle the conversion from fastq to fasta</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fastqc.html">fastqc()</a></code> </p>
        </td>
        <td><p>A function to handle a docker containier executing fastqc</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filterCounts.html">filterCounts()</a></code> </p>
        </td>
        <td><p>Filter a count table using a table of DE from wrapperDeseq2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gatkDNA.html">gatkDNA()</a></code> </p>
        </td>
        <td><p>Running realignment and recalibration, GATK</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getInfo.html">getInfo()</a></code> </p>
        </td>
        <td><p>Downloading the full set of genes information from HGNC</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="hashclone.html">hashclone()</a></code> </p>
        </td>
        <td><p>HashClone running A function to handle a hashclone docker container NOT READY for STABLE RELEASE, YET</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="heatmapBase.html">heatmapBase()</a></code> </p>
        </td>
        <td><p>hfc, heatmap for clustering</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="heatmaply.html">heatmaply()</a></code> </p>
        </td>
        <td><p>Function to produce an interactive heatmap using plot.ly</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="kraken.html">kraken()</a></code> </p>
        </td>
        <td><p>A function executing kraken</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="macs2.html">macs2()</a></code> </p>
        </td>
        <td><p>A function to handle a MACS2 containier</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mergeData.html">mergeData()</a></code> </p>
        </td>
        <td><p>Function to merge different circRNA lists from CIRI 2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mirnaCounts.html">mirnaCounts()</a></code> </p>
        </td>
        <td><p>Counting miRNAs, Cordero et al. PLoS One. 2012;7:e31630</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mirnaCovar.html">mirnaCovar()</a></code> </p>
        </td>
        <td><p>Adding covariates and batch information to miRNAseq raw counts output</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="multiQC.html">multiQC()</a></code> </p>
        </td>
        <td><p>A function to handle a docker containier executing MultiQC</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="oncosnp.html">oncosnp()</a></code> </p>
        </td>
        <td><p>A function to handle a docker containier executing CNV calculation. Only supporting hg19</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="oncosnpAnnotation.html">oncosnpAnnotation()</a></code> </p>
        </td>
        <td><p>A function to associate CNV generated by oncSNP starting from log R ratio and B allel frequences, generated by SNP arrays, to gene symbols</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pca.html">pca()</a></code> </p>
        </td>
        <td><p>generating a PCA from counts, FPKM and TPM tables from rnaseqCounts outuputs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="platypus.html">platypus()</a></code> </p>
        </td>
        <td><p>Platypus analysis NOT READY TO GO ON STABLE missing test set</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="platypusFilter.html">platypusFilter()</a></code> </p>
        </td>
        <td><p>Platypus filter analysis NOT READY TO GO ON STABLE to be used in case the full platypus workflow fail.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="prepare4expands.html">prepare4expands()</a></code> </p>
        </td>
        <td><p>A function to handle a docker containier preparing the file needed for expanse</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pubmedNet.html">pubmedNet()</a></code> </p>
        </td>
        <td><p>Creating networks on the basis of the links between genes ids and pubmedid</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rnaseqCounts.html">rnaseqCounts()</a></code> </p>
        </td>
        <td><p>Running RNAseq counting workflow for a single sample</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rsemBw.html">rsemBw()</a></code> </p>
        </td>
        <td><p>Creating a bigwig using RSEM</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rsemanno.html">rsemanno()</a></code> </p>
        </td>
        <td><p>Annotating RSEM gene.results using ENSEMBL annotation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rsemannoByGtf.html">rsemannoByGtf()</a></code> </p>
        </td>
        <td><p>Annotating RSEM gene.results using ENSEMBL gtf and refGenome CRAN package</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rsemstar.html">rsemstar()</a></code> </p>
        </td>
        <td><p>Running RSEM, Li and Dewey BMC Bioinformatics 2011 12:323</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rsemstarIndex.html">rsemstarIndex()</a></code> </p>
        </td>
        <td><p>Generating rsem-star genome index</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rsemstarUscsIndex.html">rsemstarUscsIndex()</a></code> </p>
        </td>
        <td><p>Generating rsem-star genome index</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runDocker.html">runDocker()</a></code> </p>
        </td>
        <td><p>Run docker container</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="salmonAnnotation.html">salmonAnnotation()</a></code> </p>
        </td>
        <td><p>A function to annotate salmon output</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="salmonCounts.html">salmonCounts()</a></code> </p>
        </td>
        <td><p>A function to handle a salmon docker container</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="salmonIndex.html">salmonIndex()</a></code> </p>
        </td>
        <td><p>A function to create a Salmon pseudo reference</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sample2experiment.html">sample2experiment()</a></code> </p>
        </td>
        <td><p>generating counts, FPKM and TPM tables from rnaseqCounts outuputs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sampleSize.html">sampleSize()</a></code> </p>
        </td>
        <td><p>A wrapper function for sample_size_distribution function from RnaSeqSampleSize Bioconductor package</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="skeleton.html">skeleton()</a></code> </p>
        </td>
        <td><p>A skeleton function to handle a docker containier</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="skewer.html">skewer()</a></code> </p>
        </td>
        <td><p>Running skewer, an adapter trimmer application, Jiang et al BMC Bioinformatics201415:182</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="skewerXSergio.html">skewerXSergio()</a></code> </p>
        </td>
        <td><p>Running skewer, an adapter trimmer application, Jiang et al BMC Bioinformatics201415:182</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sncRNA.html">sncRNA()</a></code> </p>
        </td>
        <td><p>Running small RNA-seq single-end reads alignment and quantification using BWA and custom scripts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="snvPreprocessing.html">snvPreprocessing()</a></code> </p>
        </td>
        <td><p>Running indel relignment and quality recalibration on bam generated by bwa function NOT READY TO GO ON STABLE</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sraDownload.html">sraDownload()</a></code> </p>
        </td>
        <td><p>A function to handle fasterq-dumper SRA to download SRA fastq files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="star2steps.html">star2steps()</a></code> </p>
        </td>
        <td><p>Running Star two steps for variant calls</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="starChimeric.html">starChimeric()</a></code> </p>
        </td>
        <td><p>Running Star to detect chimeric transcripts on paired-end sequences</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="starChipIndex.html">starChipIndex()</a></code> </p>
        </td>
        <td><p>Preparing the bed file required by starchip to detect circular RNAs on paired-end sequences</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="starchipCircle.html">starchipCircle()</a></code> </p>
        </td>
        <td><p>Running starchip to detect circular RNAs on paired-end sequences</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="testSeqbox.html">testSeqbox()</a></code> </p>
        </td>
        <td><p>Testing the SeqBox basic installation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="trimFasta.html">trimFasta()</a></code> </p>
        </td>
        <td><p>A function to execute seqtk trimming</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="wrapperCiri.html">wrapperCiri()</a></code> </p>
        </td>
        <td><p>Wrapper function for circRNAs prediction using CIRI 2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="wrapperDeseq2.html">wrapperDeseq2()</a></code> </p>
        </td>
        <td><p>A wrapper function for deseq2 for two groups only</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="wrapperPdx.html">wrapperPdx()</a></code> </p>
        </td>
        <td><p>Running PDX data preprocessing TO BE REVISED</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="wrapperSTARChip.html">wrapperSTARChip()</a></code> </p>
        </td>
        <td><p>Wrapper function for circRNAs prediction using STARChip</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="wrapperSalmon.html">wrapperSalmon()</a></code> </p>
        </td>
        <td><p>A function to estimate counts using Salmon quasi-alignment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="xenome.html">xenome()</a></code> </p>
        </td>
        <td><p>Running xenome, https://github.com/data61/gossamer/</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="xenomeIndex.html">xenomeIndex()</a></code> </p>
        </td>
        <td><p>Generating xenome genome indexes</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-all-functions">All functions</a></li>
    </ul>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Raffaele Calogero, Neha Kulkarni, Luca Alessandri, Marco Beccuti, Greta Romano, Francesca Cordero.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


