<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>A function to handle a MACS2 containier — macs2 • docker4seq</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="A function to handle a MACS2 containier — macs2"><meta property="og:description" content="This function executes a MACS2 docker that produces as output peaks call"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">docker4seq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.4.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>A function to handle a MACS2 containier</h1>
    
    <div class="hidden name"><code>macs2.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function executes a MACS2 docker that produces as output peaks call</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">macs2</span><span class="op">(</span></span>
<span>  group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sudo"</span>, <span class="st">"docker"</span><span class="op">)</span>,</span>
<span>  <span class="va">control.bam</span>,</span>
<span>  <span class="va">chipseq.bam</span>,</span>
<span>  <span class="va">experiment.name</span>,</span>
<span>  histone.marks <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  broad.cutoff <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  qvalue <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  organism <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hs"</span>, <span class="st">"mm"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>group, </dt>
<dd><p>a character string. Two options: sudo or docker, depending to which group the user belongs</p></dd>


<dt>control.bam, </dt>
<dd><p>a character string indicating the path to the control bam file. IMPORTANT control.bam and chipseq.bam are in the same folder</p></dd>


<dt>chipseq.bam, </dt>
<dd><p>a character string indicating the path to the chipseq bam file. IMPORTANT control.bam and chipseq.bam are in the same folder</p></dd>


<dt>experiment.name, </dt>
<dd><p>a character string indicating the prefix for MCS2 output</p></dd>


<dt>histone.marks, </dt>
<dd><p>boolean if TRUE activate the broad option to call histone marks</p></dd>


<dt>broad.cutoff, </dt>
<dd><p>if histone.mark is TRUE broad.cutoff can be set, default 0.1</p></dd>


<dt>qvalue, </dt>
<dd><p>The qvalue (minimum FDR) cutoff to call significant regions. Default is 0.05. For broad marks, you can try 0.05 as cutoff.</p></dd>


<dt>organism, </dt>
<dd><p>required to select the correct genome size avaialble options hs, mm</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>NAME_peaks.xls, which is a tabular file which contains information about called peaks. You can open it in excel and sort/filter using excel functions. Information include: chromosome name, start position of peak, end position of peak, length of peak region, absolute peak summit position, pileup height at peak summit, -log10(pvalue) for the peak summit (e.g. pvalue =1e-10, then this value should be 10), fold enrichment for this peak summit against random Poisson distribution with local lambda, -log10(qvalue) at peak summit. NAME_peaks.narrowPeak is BED6+4 format file which contains the peak locations together with peak summit, pvalue and qvalue. You can load it to UCSC genome browser. Definition of some specific columns are: 5th: integer score for display calculated as int(-10*log10qvalue). Please note that currently this value might be out of the [0-1000] range defined in UCSC Encode narrowPeak format, 7th: fold-change, 8th: -log10pvalue, 9th: -log10qvalue, 10th: relative summit position to peak start. NAME_peaks.broadPeak is in BED6+3 format which is similar to narrowPeak file.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>raffaele.calogero [at] unito [dot] it, University of Torino</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>    <span class="co">#running MACS for conventional peaks</span></span></span>
<span class="r-in"><span>    <span class="fu">macs2</span><span class="op">(</span>group<span class="op">=</span><span class="st">"docker"</span>, control.bam<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>,<span class="st">"igg_dedup_reads.bam"</span>, sep<span class="op">=</span><span class="st">"/"</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>            chipseq.bam<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>,<span class="st">"prdm51_dedup_reads.bam"</span>, sep<span class="op">=</span><span class="st">"/"</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>            experiment.name<span class="op">=</span><span class="st">"prdm51_igg"</span>, histone.marks<span class="op">=</span><span class="cn">FALSE</span>, qvalue<span class="op">=</span><span class="fl">0.05</span>, </span></span>
<span class="r-in"><span>            organism<span class="op">=</span><span class="st">"mm"</span><span class="op">)</span></span></span>
<span class="r-in"><span>            </span></span>
<span class="r-in"><span>    <span class="co">#running MACS for histone marks</span></span></span>
<span class="r-in"><span>    <span class="fu">macs2</span><span class="op">(</span>group<span class="op">=</span><span class="st">"docker"</span>, control.bam<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>,<span class="st">"igg_dedup_reads.bam"</span>, sep<span class="op">=</span><span class="st">"/"</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>            chipseq.bam<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>,<span class="st">"prdm51_dedup_reads.bam"</span>, sep<span class="op">=</span><span class="st">"/"</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>            experiment.name<span class="op">=</span><span class="st">"prdm51_igg"</span>, histone.marks<span class="op">=</span><span class="cn">TRUE</span>, broad.cutoff<span class="op">=</span><span class="fl">0.1</span>, </span></span>
<span class="r-in"><span>            organism<span class="op">=</span><span class="st">"mm"</span><span class="op">)</span>     </span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Raffaele Calogero, Neha Kulkarni, Luca Alessandri, Marco Beccuti, Greta Romano, Francesca Cordero.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer></div>

  


  

  </body></html>

