<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Platypus filter analysis NOT READY TO GO ON STABLE to be used in case the full platypus workflow fail. — platypusFilter • docker4seq</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Platypus filter analysis NOT READY TO GO ON STABLE to be used in case the full platypus workflow fail. — platypusFilter" />
<meta property="og:description" content="This function executes platypusFilter filtering the single_variants.vcf generated by platypys function." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">docker4seq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/docker4seq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/skeleton.html">Skeleton tutorial</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Platypus filter analysis NOT READY TO GO ON STABLE to be used in case the full platypus workflow fail.</h1>
    
    <div class="hidden name"><code>platypusFilter.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function executes platypusFilter filtering the single_variants.vcf generated by platypys function.</p>
    </div>

    <pre class="usage"><span class='fu'>platypusFilter</span>(
  <span class='kw'>group</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"sudo"</span>, <span class='st'>"docker"</span>),
  <span class='kw'>data.folder</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/getwd.html'>getwd</a></span>(),
  <span class='no'>scratch.folder</span>,
  <span class='no'>GQ</span>,
  <span class='no'>minSampGQ</span>,
  <span class='no'>NR</span>,
  <span class='no'>minSampNR</span>,
  <span class='no'>NV</span>,
  <span class='no'>minSampNV</span>,
  <span class='no'>normal_samples</span>,
  <span class='no'>GT_normal</span>,
  <span class='no'>minSampGT_normal</span>,
  <span class='no'>tumoral_samples</span>,
  <span class='no'>GT_tumoral</span>,
  <span class='no'>minSampGT_tumoral</span>,
  <span class='no'>stringent_filter</span>,
  <span class='no'>annotation</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>group, </th>
      <td><p>a character string. Two options: sudo or docker, depending to which group the user belongs</p></td>
    </tr>
    <tr>
      <th>data.folder, </th>
      <td><p>a character string indicating the folder where bams and vcf files are located and where output will be written</p></td>
    </tr>
    <tr>
      <th>scratch.folder, </th>
      <td><p>a character string indicating the path of the scratch folder</p></td>
    </tr>
    <tr>
      <th>GQ, </th>
      <td><p>min GQ value to consider (extreme included)</p></td>
    </tr>
    <tr>
      <th>minSampGQ, </th>
      <td><p>min number of samples with GQ value (extreme included), usually we start with 85% of the samples</p></td>
    </tr>
    <tr>
      <th>NR, </th>
      <td><p>min NR value to consider (extreme included), the number of reads covering the SNV region</p></td>
    </tr>
    <tr>
      <th>minSampNR, </th>
      <td><p>min number of samples with NR value (extreme included), usually we start with 85% of the samples</p></td>
    </tr>
    <tr>
      <th>NV, </th>
      <td><p>min NV value to consider (extreme included), the n of reads with the SNV</p></td>
    </tr>
    <tr>
      <th>minSampNV, </th>
      <td><p>min number of samples with NV value (extreme included)</p></td>
    </tr>
    <tr>
      <th>normal_samples, </th>
      <td><p>string with names (group names of bam files) of normal samples separated by hash, write NULL if you do not want to use GT filter in normal samples</p></td>
    </tr>
    <tr>
      <th>GT_normal, </th>
      <td><p>GT value in normal samples to consider, type "NO" if you do not want to use this filter, else e.g. you might use 0/0</p></td>
    </tr>
    <tr>
      <th>minSampGT_normal, </th>
      <td><p>min number of normal samples with GT value</p></td>
    </tr>
    <tr>
      <th>tumoral_samples, </th>
      <td><p>string with names (group names of bam files) of tumoral samples separated by &amp;, write NULL if you do not want to use GT filter in normal samples</p></td>
    </tr>
    <tr>
      <th>GT_tumoral, </th>
      <td><p>GT value in tumoral samples to consider, type "NO" if you do not want to use this filter. Type e.g. 0/0 if you do not want to consider this genotype.</p></td>
    </tr>
    <tr>
      <th>minSampGT_tumoral, </th>
      <td><p>min number of tumoral samples in which the GT value is NOT present</p></td>
    </tr>
    <tr>
      <th>stringent_filter, </th>
      <td><p>To enable the filter (it keeps only the variants with "PASS" value or the variants that have only the "alleleBias" value) use 1 or 0 to disable it</p></td>
    </tr>
    <tr>
      <th>annotation, </th>
      <td><p>hg19 and mm10 are actually available for the annotation of the detected SNVs</p></td>
    </tr>
    </table>


    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
    <span class='co'>#filtering platypus results </span>
 <span class='fu'>platypusFilter</span>(<span class='kw'>group</span><span class='kw'>=</span><span class='st'>"docker"</span>, <span class='kw'>data.folder</span><span class='kw'>=</span><span class='st'>"/archive/home/rcaloger/data/platypus_tests/mm10ENSEMBL"</span>,
        <span class='kw'>scratch.folder</span><span class='kw'>=</span><span class='st'>"/scratch/users/rcaloger/"</span>,
        <span class='kw'>GQ</span><span class='kw'>=</span><span class='fl'>10</span>, <span class='kw'>minSampGQ</span><span class='kw'>=</span><span class='fl'>2</span>, <span class='kw'>NR</span><span class='kw'>=</span><span class='fl'>10</span>, <span class='kw'>minSampNR</span><span class='kw'>=</span><span class='fl'>2</span>, <span class='kw'>NV</span><span class='kw'>=</span><span class='fl'>3</span>,
        <span class='kw'>minSampNV</span><span class='kw'>=</span><span class='fl'>1</span>, <span class='kw'>normal_samples</span><span class='kw'>=</span><span class='st'>"MAMBO43"</span>, <span class='kw'>GT_normal</span><span class='kw'>=</span><span class='st'>"0/0"</span>,
        <span class='kw'>minSampGT_normal</span><span class='kw'>=</span><span class='fl'>1</span>, <span class='kw'>tumoral_samples</span><span class='kw'>=</span><span class='st'>"MAMBO43TRT#MAMBO43TRNT"</span>, <span class='kw'>GT_tumoral</span><span class='kw'>=</span><span class='st'>"0/0"</span>,
        <span class='kw'>minSampGT_tumoral</span><span class='kw'>=</span><span class='fl'>1</span>, <span class='kw'>stringent_filter</span><span class='kw'>=</span><span class='fl'>0</span>, <span class='kw'>annotation</span><span class='kw'>=</span><span class='st'>"mm10"</span>)
}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Riccardo Panero, riccardo.panero[at]gmail[dot]com, Bioinformatics and Genomics unit, University of Torino Italy</p>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Raffaele Calogero, Neha Kulkarni, Luca Alessandri, Marco Beccuti, Greta Romano, Francesca Cordero.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


