<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Running RNAseq counting workflow for a single sample — rnaseqCounts • docker4seq</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Running RNAseq counting workflow for a single sample — rnaseqCounts"><meta property="og:description" content="This function executes a set of docker containers allowing the generation of gene and isoforms counts for a single sample.
#params skewer"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">docker4seq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.4.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Running RNAseq counting workflow for a single sample</h1>
    
    <div class="hidden name"><code>rnaseqCounts.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function executes a set of docker containers allowing the generation of gene and isoforms counts for a single sample.
#params skewer</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">rnaseqCounts</span><span class="op">(</span></span>
<span>  group <span class="op">=</span> <span class="st">"sudo"</span>,</span>
<span>  fastq.folder <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  scratch.folder <span class="op">=</span> <span class="st">"/data/scratch"</span>,</span>
<span>  threads <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  <span class="va">adapter5</span>,</span>
<span>  <span class="va">adapter3</span>,</span>
<span>  seq.type <span class="op">=</span> <span class="st">"pe"</span>,</span>
<span>  min.length <span class="op">=</span> <span class="fl">40</span>,</span>
<span>  genome.folder <span class="op">=</span> <span class="st">"/data/genomes/hg38star"</span>,</span>
<span>  strandness <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  save.bam <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  org <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>  annotation.type <span class="op">=</span> <span class="st">"gtfENSEMBL"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>group, </dt>
<dd><p>a character string. Two options: <code>"sudo"</code> or <code>"docker"</code>, depending to which group the user belongs</p></dd>


<dt>fastq.folder, </dt>
<dd><p>a character string indicating where gzip fastq files are located</p></dd>


<dt>scratch.folder, </dt>
<dd><p>a character string indicating the scratch folder where docker container will be mounted</p></dd>


<dt>threads, </dt>
<dd><p>a number indicating the number of cores to be used from the application</p></dd>


<dt>adapter5, </dt>
<dd><p>a character string indicating the fwd adapter</p></dd>


<dt>adapter3, </dt>
<dd><p>a character string indicating the rev adapter</p></dd>


<dt>seq.type, </dt>
<dd><p>a character string indicating the type of reads to be generated by the sequencer. Two options: <code>"se"</code> or <code>"pe"</code> respectively for single end and pair end sequencing.</p></dd>


<dt>min.length, </dt>
<dd><p>a number indicating minimal length required to return a trimmed read
#params rsemstar</p></dd>


<dt>genome.folder, </dt>
<dd><p>a character string indicating the folder where the indexed reference genome is located. IMPORTANT the present function only suport genomic indexes made using ensembl genom and the corresponding gtf</p></dd>


<dt>strandness, </dt>
<dd><p>a character string indicating the type ofsequencing protocol used for the analysis. Three options: <code>"none"</code>, <code>"forward"</code>, <code>"reverse"</code> respectively for non strand selection, reverse for Illumina strandness protocols, reverse for ACCESS Illumina protocol</p></dd>


<dt>save.bam, </dt>
<dd><p>a boolean value, TRUE or FALSE, to save also BAM files generated by STAR and RSEM
#params rsemanno</p></dd>


<dt>org, </dt>
<dd><p>a character string indicating the genome assembly used for mapping and counting with <code>"rsemstar"</code> function only required for biocENSEMBL based annotation</p></dd>


<dt>annotation.type, </dt>
<dd><p>a character string. Two options: <code>"biocENSEMBL"</code> or <code>"gtfENSEMBL"</code>. <code>"biocENSEMBL"</code> will annotate by Bioconductor only protein coding genes. <code>"gtfENSEMBL"</code> will annotate all RNAs described in <code>"annotation.type"</code></p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>Returns the output of skewer, rsemstar, rsemannos' functions</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Raffaele Calogero</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"wget http://130.192.119.59/public/test_R1.fastq.gz"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">docker4seq</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">rnaseqCounts</span><span class="op">(</span>group<span class="op">=</span><span class="st">"docker"</span>,fastq.folder<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>, scratch.folder<span class="op">=</span><span class="st">"/data/scratch/"</span>,</span></span>
<span class="r-in"><span>            adapter5<span class="op">=</span><span class="st">"AGATCGGAAGAGCACACGTCTGAACTCCAGTCA"</span>,</span></span>
<span class="r-in"><span>            adapter3<span class="op">=</span><span class="st">"AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT"</span>,</span></span>
<span class="r-in"><span>            seq.type<span class="op">=</span><span class="st">"se"</span>, threads<span class="op">=</span><span class="fl">24</span>,  min.length<span class="op">=</span><span class="fl">40</span>,</span></span>
<span class="r-in"><span>            genome.folder<span class="op">=</span><span class="st">"/data/genomes/hg38star"</span>, strandness<span class="op">=</span><span class="st">"none"</span>, save.bam<span class="op">=</span><span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>            org<span class="op">=</span><span class="st">"hg38"</span>, annotation.type<span class="op">=</span><span class="st">"gtfENSEMBL"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Raffaele Calogero, Neha Kulkarni, Luca Alessandri, Marco Beccuti, Greta Romano, Francesca Cordero.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer></div>

  


  

  </body></html>

